src/metatensor/models/utils/data/combine_dataloaders.py

```diff
@@ -1,57 +1,58 @@
     """
-    Combines multiple dataloaders into a single iterable dataset.
-    This is useful for combining multiple dataloaders into a single
-    dataloader. The new dataloader can be shuffled or not.
-
-    :param dataloaders: list of dataloaders to combine
-    :param shuffle: whether to shuffle the combined dataloader
+    Combines multiple dataloaders into a single dataloader.
 
-    :return: combined dataloader
+    This is useful for learning from multiple datasets at the same time,
+    each of which may have different batch sizes, properties, etc.
     """
 
-    def __init__(self, dataloaders, shuffle):
+    def __init__(self, dataloaders: List[torch.utils.data.DataLoader], shuffle: bool):
+        """Creates the combined dataloader.
+
+        :param dataloaders: list of dataloaders to combine
+        :param shuffle: whether to shuffle the combined dataloader (this does not
+            act on the individual batches, but it shuffles the order in which
+            they are returned)
+
+        :return: the combined dataloader
+        """
+
         self.dataloaders = dataloaders
         self.shuffle = shuffle
 
         # Create the indices:
-        indices = [
-            (i, dl_idx)
-            for dl_idx, dl in enumerate(self.dataloaders)
-            for i in range(len(dl))
-        ]
+        self.indices = list(range(len(self)))
 
         # Shuffle the indices if requested
         if self.shuffle:
-            np.random.shuffle(indices)
+            np.random.shuffle(self.indices)
 
-        self.indices = indices
+        self.reset()
 
-    def __iter__(self):
-        for idx, dataloader_idx in self.indices:
-            yield next(itertools.islice(self.dataloaders[dataloader_idx], idx, None))
-
-    def __len__(self):
-        return len(self.indices)
+    def reset(self):
+        self.current_index = 0
+        self.full_list = [batch for dl in self.dataloaders for batch in dl]
 
+    def __iter__(self):
+        return self
 
-def combine_dataloaders(
-    dataloaders: List[torch.utils.data.DataLoader], shuffle: bool = True
-):
-    """
-    Combines multiple dataloaders into a single dataloader.
+    def __next__(self):
+        if self.current_index >= len(self.indices):
+            self.reset()  # Reset the index for the next iteration
+            raise StopIteration
 
-    :param dataloaders: list of dataloaders to combine
-    :param shuffle: whether to shuffle the combined dataloader
+        idx = self.indices[self.current_index]
+        self.current_index += 1
+        return self.full_list[idx]
 
-    :return: combined dataloader
-    """
-    combined_dataset = CombinedIterableDataset(dataloaders, shuffle)
-    return torch.utils.data.DataLoader(combined_dataset, batch_size=None)
+    def __len__(self):
+        # this returns the total number of batches in all dataloaders
+        # (as opposed to the total number of samples or the number of
+        # individual dataloaders)
```

Comment:
Maybe this should be visible mostly users -> move to docstrings

Once I am reading the code I don't need this information because the lines below already tells me this.